### ISSUE TITLE
Confused about the implementation of equation 3 in the paper

### ISSUE THREAD

**Issue by Tcorpion** [lioryariv/idr#44]
#### equation 3  in paper

![00](https://github.com/lioryariv/idr/assets/19811908/c1e36bdc-7976-4da8-b052-238e07f4f3d0)

#### the code in this repo

![01](https://github.com/lioryariv/idr/assets/19811908/a8ef4a31-8aef-4acd-a2d5-d0b1f7e7c637)

Compared  the implementation of the description of equation 3 to the repo code, I am confused about the code `surface_dists_theta = surface_dists - (surface_output - surface_sdf_values) / surface_points_dot`, here `surface_output` is `f (c + t 0 v; θ)` and is equal to `surface_sdf_values` which has a detach operation in `surface_sdf_values = output[:N, 0:1].detach()`,  it seems that `f (c + t 0 v; θ) - f (c + t 0 v; θ).detach()`,  should the code consistent with the formula be `surface_dists_theta = surface_dists - surface_output  / surface_points_dot`?

**Comment by ArtoriasAbyssslayer** [comment]
_Disclaimer: Sorry for making wrong assumptions. This is my idea on the topic only._ 

Making a comment on this I think the procedure is used, because you try to train a Neural Level Set which represents a differentiable manifold using a sample network linear layer(to control it) so by detaching the previous values from the computation graph you now focus on the points that were "sampled" and didn't converge(and the difference btween these two). Detach function is mainly used to stop calculating gradients as it detaches the tensor from the neural network(pytorch wont calculate gradients). I think the surface_sdf_values are detached because they are not going to be used any more and the return differentiable surface points are the ones which geometry parameters should be trained. So in this way you train the difference of the sampled points to the previous points inside the sample network.  Also this is done only during training and if there are intersection points N>1. In inference the sdf_ouput(trained) represents the isometric surface - manifold (neural mask is used to take it from points)  implicitly as a neural level set and  when the training finishes it should have trained geometry. I hope this is clear using the surface_output only wont result in any training of the distance of the sample points and the points that already belong to sdf implicit surface. I think of it as a discreet differntial equation and that $t_0$ on the equation is meaningful. 

_I tend to ramble(wrote that fast too) and may make errors in my logic, but this is my perspective. Corrections are welcome_

**Comment by Tcorpion** [comment]
> _Disclaimer: Sorry for making wrong assumptions. This is my idea on the topic only._
> 
> Making a comment on this I think the procedure is used, because you try to train a Neural Level Set which represents a differentiable manifold using a sample network linear layer(to control it) so by detaching the previous values from the computation graph you now focus on the points that were "sampled" and didn't converge(and the difference btween these two). Detach function is mainly used to stop calculating gradients as it detaches the tensor from the neural network(pytorch wont calculate gradients). I think the surface_sdf_values are detached because they are not going to be used any more and the return differentiable surface points are the ones which geometry parameters should be trained. So in this way you train the difference of the sampled points to the previous points inside the sample network. Also this is done only during training and if there are intersection points N>1. In inference the sdf_ouput(trained) represents the isometric surface - manifold (neural mask is used to take it from points) implicitly as a neural level set and when the training finishes it should have trained geometry. I hope this is clear using the surface_output only wont result in any training of the distance of the sample points and the points that already belong to sdf implicit surface. I think of it as a discreet differntial equation and that t0 on the equation is meaningful.
> 
> _I tend to ramble(wrote that fast too) and may make errors in my logic, but this is my perspective. Corrections are welcome_

`surface_dists_theta = surface_dists - surface_output  / surface_points_dot` also represents the differentiable manifold,  `(surface_output - surface_sdf_values)` gives zero value,   `surface_output` stands close to zero, why did not the authors use  `surface_dists_theta = surface_dists - surface_output  / surface_points_dot`?

**Comment by ArtoriasAbyssslayer** [comment]
Since I don't think the point I wanted to make above was conveyed, I refer you to read the previous paper οn [Controlling neural level sets](https://arxiv.org/abs/1905.11911)  and specifically the paragraph 2.1and 2.2 that talks about Level Set Sampling and Differentiable sample position

**Comment by Tcorpion** [comment]
> Since I don't think the point I wanted to make above was conveyed, I refer you to read the previous paper οn [Controlling neural level sets](https://arxiv.org/abs/1905.11911) and specifically the paragraph 2.1 that talks about Level Set Sampling and the Sample Network

Thanks
